custom_components/kia_uvo/button.py

```diff
@@ -0,0 +1,215 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable
+
+from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+
+
+@dataclass(frozen=True, kw_only=True)
+class KiaButtonDescription(ButtonEntityDescription):
+    """A class that describes custom button entities, including service, payload, and capability metadata."""
+
+    service: str
+    payload: dict[str, Any] | None = None
+    capability: Callable[[Any, str], bool] | None = (
+        None  # (coordinator, vehicle_id) -> bool
+    )
+
+
+def _is_ev(coordinator, vehicle_id: str) -> bool:
+    """Return True if this vehicle is an EV (robust, attribute-based)."""
+    vm = getattr(coordinator, "vehicle_manager", None)
+    vehicles = getattr(vm, "vehicles", {}) if vm else {}
+    v = vehicles.get(vehicle_id)
+    if v is None:
+        return False
+
+    if getattr(v, "ev_charge_limits_ac", None) is not None:
+        return True
+    if getattr(v, "ev_charge_limits_dc", None) is not None:
+        return True
+
+    return False
+
+
+BUTTONS: list[KiaButtonDescription] = [
+    KiaButtonDescription(
+        key="update_cached",
+        name="Update (cached)",
+        icon="mdi:update",
+        service="update",
+    ),
+    KiaButtonDescription(
+        key="force_update",
+        name="Force update",
+        icon="mdi:cloud-refresh",
+        service="force_update",
+    ),
+    KiaButtonDescription(
+        key="lock",
+        name="Lock",
+        icon="mdi:lock",
+        service="lock",
+    ),
+    KiaButtonDescription(
+        key="unlock",
+        name="Unlock",
+        icon="mdi:lock-open",
+        service="unlock",
+    ),
+    KiaButtonDescription(
+        key="start_climate",
+        name="Start climate",
+        icon="mdi:car-seat-heater",
+        service="start_climate",
+    ),
+    KiaButtonDescription(
+        key="stop_climate",
+        name="Stop climate",
+        icon="mdi:car-seat-cooler",
+        service="stop_climate",
+    ),
+    KiaButtonDescription(
+        key="start_charge",
+        name="Start charge",
+        icon="mdi:ev-station",
+        service="start_charge",
+        capability=_is_ev,
+    ),
+    KiaButtonDescription(
+        key="stop_charge",
+        name="Stop charge",
+        icon="mdi:ev-station-off",
+        service="stop_charge",
+        capability=_is_ev,
```

Comment:
maybe a spot to use a lambda for the vehicle.engine_type?

Reply:
using this now: `capability=lambda v: _is_electric(v),`